src/filter/mod.rs

```diff
@@ -321,54 +341,72 @@ where
 /// Returns 2d correlation of an image with the outer product of the 1d
 /// kernel filter with itself.
 #[must_use = "the function does not modify the original image"]
-pub fn separable_filter_equal<P, K>(image: &Image<P>, kernel: &[K]) -> Image<P>
+pub fn separable_filter_equal<P, K>(image: &Image<P>, kernel: &Kernel<K>) -> Image<P>
 where
     P: Pixel,
     <P as Pixel>::Subpixel: Into<K> + Clamp<K>,
     K: Num + Copy,
 {
+    assert_eq!(
+        kernel.height, 1,
+        "separable_filter only works with 1 dimensional kernels"
+    );
+
     separable_filter(image, kernel, kernel)
 }
 
 /// Returns 2d correlation of an image with a 3x3 row-major kernel. Intermediate calculations are
 /// performed at type K, and the results clamped to subpixel type S. Pads by continuity.
 #[must_use = "the function does not modify the original image"]
-pub fn filter3x3<P, K, S>(image: &Image<P>, kernel: &[K]) -> Image<ChannelMap<P, S>>
+pub fn filter3x3<P, K, S>(image: &Image<P>, kernel: &Kernel<K>) -> Image<ChannelMap<P, S>>
```

Comment:
When this function was originally written it used a handwritten loop that I hoped (and maybe benchmarked? I don’t remember - it’s been a long time!) would be faster than a generic nxm kernel calculation. However, now that this function just calls the regular `filter` function I’m tempted to just bin it - there’s no longer anything special about 3x3.

Reply:
I've been quite surprised when going through the code at the amount or manual loop unrolling and so on that makes the code much harder to reason about. I understand this was originally done for performance reasons but I don't think writing code in this intentionally obtuse way is worth the cost to readability and maintenance. It seem like a bit of premature optimization since this library has yet to really stabilize much of its API.

I agree with removing `filter3x3()`. While on the topic is `filter()` even the right name when the operation is actually a convolution? Should we rename `filter()` to `convolve()` while we're at it?